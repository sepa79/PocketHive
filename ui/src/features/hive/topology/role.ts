export type Role =
  | 'Seeder'
  | 'Gatekeeper'
  | 'Worker'
  | 'Postprocessor'
  | 'Trigger'
  | 'LogAggregator';

export const BeeTitle: Record<Role, string> = {
  Seeder: 'Seeder',
  Gatekeeper: 'Gatekeeper',
  Worker: 'Worker',
  Postprocessor: 'BeeCounter',
  Trigger: 'Activator',
  LogAggregator: 'Scribe',
};
