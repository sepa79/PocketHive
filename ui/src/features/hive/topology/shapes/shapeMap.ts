import type { Role } from '../role';

export const roleShape: Record<Role, string> = {
  Seeder: 'hexagon',
  Gatekeeper: 'shield',
  Worker: 'roundedRect',
  Postprocessor: 'cylinder',
  Trigger: 'triangleRight',
  LogAggregator: 'documentFolded',
};
