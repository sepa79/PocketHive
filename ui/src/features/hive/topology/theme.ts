export const BeeColors = {
  nodeFill: 'var(--ph-node-fill)',
  nodeStroke: 'var(--ph-node-stroke)',
  text: 'var(--ph-text)',
  status: {
    ok: 'var(--ph-ok)',
    warn: 'var(--ph-warn)',
    err: 'var(--ph-err)',
    ghost: 'var(--ph-ghost)',
  },
} as const;

export type Status = keyof typeof BeeColors.status;

export const BeeSizes = {
  w: 140,
  h: 72,
  radius: 10,
  halo: 8,
} as const;

export function statusToColor(status: Status): string {
  return BeeColors.status[status];
}
