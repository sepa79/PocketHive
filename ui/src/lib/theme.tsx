import { ReactNode, createContext, useContext, useEffect, useMemo, useState } from 'react'

export type Theme = 'dark' | 'light'

type ThemeContextValue = {
  theme: Theme
  setTheme: (theme: Theme) => void
}

const ThemeContext = createContext<ThemeContextValue>({
  theme: 'dark',
  setTheme: () => undefined,
})

export type ThemeProviderProps = {
  children: ReactNode
  initialTheme?: Theme
}

export function ThemeProvider({ children, initialTheme = 'dark' }: ThemeProviderProps) {
  const [theme, setTheme] = useState<Theme>(initialTheme)

  useEffect(() => {
    if (typeof document !== 'undefined') {
      document.documentElement.dataset.theme = theme
    }
  }, [theme])

  const value = useMemo(() => ({ theme, setTheme }), [theme])

  return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
}

export function useTheme() {
  return useContext(ThemeContext)
}
