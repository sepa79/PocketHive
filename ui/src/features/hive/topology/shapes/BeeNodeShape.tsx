import type { FC } from 'react';
import { BeeColors, BeeSizes, type Status } from '../theme';
import { hexagon, shield, roundedRect, triangleRight, documentFolded, cylinder } from './paths';
import type { Role } from '../role';

export interface NodeVisualProps {
  role: Role;
  status: Status;
  title: string;
  name: string;
  w?: number;
  h?: number;
  selected?: boolean;
  progress?: number; // for Postprocessor
}

export const BeeNodeShape: FC<NodeVisualProps> = ({
  role,
  status,
  title,
  name,
  w = BeeSizes.w,
  h = BeeSizes.h,
  selected = false,
  progress = 0,
}) => {
  const stroke = BeeColors.nodeStroke;
  const fill = BeeColors.nodeFill;
  const halo = BeeColors.status[status];

  const shape = (() => {
    switch (role) {
      case 'Seeder':
        return <path d={hexagon(w, h)} fill={fill} stroke={stroke} />;
      case 'Gatekeeper':
        return <path d={shield(w, h)} fill={fill} stroke={stroke} />;
      case 'Worker':
        return (
          <>
            <path d={roundedRect(w, h, BeeSizes.radius)} fill={fill} stroke={stroke} />
            <g opacity="0.5">
              <polyline points="-20,-6 -8,0 -20,6" fill="none" stroke={stroke} />
              <polyline points="0,-6 12,0 0,6" fill="none" stroke={stroke} />
            </g>
          </>
        );
      case 'Postprocessor': {
        const { rx, ry, yTop, yBottom } = cylinder(w, h);
        const ringR = Math.max(w, h) / 2 + BeeSizes.halo;
        const sweep = Math.max(0.01, Math.min(1, progress)) * 2 * Math.PI;
        const x1 = ringR * Math.cos(-Math.PI / 2);
        const y1 = ringR * Math.sin(-Math.PI / 2);
        const x2 = ringR * Math.cos(-Math.PI / 2 + sweep);
        const y2 = ringR * Math.sin(-Math.PI / 2 + sweep);
        const large = sweep > Math.PI ? 1 : 0;

        return (
          <>
            <ellipse cx="0" cy={yTop} rx={rx} ry={ry} fill={fill} stroke={stroke} />
            <rect x={-rx} y={yTop} width={rx * 2} height={h} fill={fill} stroke={stroke} />
            <ellipse cx="0" cy={yBottom} rx={rx} ry={ry} fill={fill} stroke={stroke} />
            <path
              d={`M 0,0 m ${x1},${y1} A ${ringR},${ringR} 0 ${large} 1 ${x2},${y2}`}
              fill="none"
              stroke={halo}
              strokeWidth="3"
            />
          </>
        );
      }
      case 'Trigger':
        return (
          <>
            <path d={triangleRight(w, h)} fill={fill} stroke={stroke} />
            <polyline points="-6,-10 4,0 -6,10" fill="none" stroke={stroke} />
          </>
        );
      case 'LogAggregator': {
        const { body, fold } = documentFolded(w, h);
        return (
          <>
            <path d={body} fill={fill} stroke={stroke} />
            <path d={fold} fill="none" stroke={stroke} />
            <g opacity="0.5">
              <line x1="-40" y1="-8" x2="30" y2="-8" stroke={stroke} />
              <line x1="-40" y1="8" x2="18" y2="8" stroke={stroke} />
            </g>
          </>
        );
      }
    }
  })();

  return (
    <g className={`bee-node ${selected ? 'is-selected' : ''}`}>
      <circle
        cx="0"
        cy="0"
        r={Math.max(w, h) / 2 + BeeSizes.halo}
        fill="none"
        stroke={halo}
        strokeOpacity={status === 'ghost' ? 0.4 : 0.8}
        strokeDasharray={status === 'ghost' ? '4 4' : '0'}
      />
      {shape}
      <g textAnchor="middle" fill={BeeColors.text}>
        <text y={-8} fontSize="12" fontWeight="700">
          {title}
        </text>
        <text y={10} fontSize="11">
          {name}
        </text>
      </g>
    </g>
  );
};
