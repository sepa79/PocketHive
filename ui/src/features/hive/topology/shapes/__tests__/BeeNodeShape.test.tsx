import { render } from '@testing-library/react';
import React from 'react';
import { describe, test, expect } from 'vitest';
import { BeeNodeShape } from '../BeeNodeShape';
import { roleShape } from '../shapeMap';
import { BeeTitle, Role } from '../../role';
import { statusToColor, BeeColors, Status } from '../../theme';

describe('role to shape mapping', () => {
  test('map contains all roles', () => {
    const roles: Role[] = ['Seeder', 'Gatekeeper', 'Worker', 'Postprocessor', 'Trigger', 'LogAggregator'];
    for (const r of roles) {
      expect(roleShape[r]).toBeDefined();
    }
  });
});

describe('statusToColor', () => {
  const statuses: Status[] = ['ok', 'warn', 'err', 'ghost'];
  for (const s of statuses) {
    test(`status ${s} maps to color`, () => {
      expect(statusToColor(s)).toBe(BeeColors.status[s]);
    });
  }
});

describe('BeeNodeShape snapshots', () => {
  const roles: Role[] = ['Seeder', 'Gatekeeper', 'Worker', 'Postprocessor', 'Trigger', 'LogAggregator'];
  for (const role of roles) {
    test(role, () => {
      const { container } = render(
        <svg>
          <BeeNodeShape role={role} status="ok" title={BeeTitle[role]} name="alpha" />
        </svg>
      );
      expect(container).toMatchSnapshot();
    });
  }
});
