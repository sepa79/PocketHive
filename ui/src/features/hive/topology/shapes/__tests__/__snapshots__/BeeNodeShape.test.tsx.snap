// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BeeNodeShape snapshots > Gatekeeper 1`] = `
<div>
  <svg>
    <g
      class="bee-node "
    >
      <circle
        cx="0"
        cy="0"
        fill="none"
        r="78"
        stroke="var(--ph-ok)"
        stroke-dasharray="0"
        stroke-opacity="0.8"
      />
      <path
        d="
    M -70,-24
    H 70
    Q 70,0 0,36
    Q -70,0 -70,-24
    Z
  "
        fill="var(--ph-node-fill)"
        stroke="var(--ph-node-stroke)"
      />
      <g
        fill="var(--ph-text)"
        text-anchor="middle"
      >
        <text
          font-size="12"
          font-weight="700"
          y="-8"
        >
          Gatekeeper
        </text>
        <text
          font-size="11"
          y="10"
        >
          alpha
        </text>
      </g>
    </g>
  </svg>
</div>
`;

exports[`BeeNodeShape snapshots > LogAggregator 1`] = `
<div>
  <svg>
    <g
      class="bee-node "
    >
      <circle
        cx="0"
        cy="0"
        fill="none"
        r="78"
        stroke="var(--ph-ok)"
        stroke-dasharray="0"
        stroke-opacity="0.8"
      />
      <path
        d="
    M -70,-36 H 54 L 70,-20 V 36 H -70 Z
  "
        fill="var(--ph-node-fill)"
        stroke="var(--ph-node-stroke)"
      />
      <path
        d="
    M 54,-36 L 54,-20 L 70,-20
  "
        fill="none"
        stroke="var(--ph-node-stroke)"
      />
      <g
        opacity="0.5"
      >
        <line
          stroke="var(--ph-node-stroke)"
          x1="-40"
          x2="30"
          y1="-8"
          y2="-8"
        />
        <line
          stroke="var(--ph-node-stroke)"
          x1="-40"
          x2="18"
          y1="8"
          y2="8"
        />
      </g>
      <g
        fill="var(--ph-text)"
        text-anchor="middle"
      >
        <text
          font-size="12"
          font-weight="700"
          y="-8"
        >
          Scribe
        </text>
        <text
          font-size="11"
          y="10"
        >
          alpha
        </text>
      </g>
    </g>
  </svg>
</div>
`;

exports[`BeeNodeShape snapshots > Postprocessor 1`] = `
<div>
  <svg>
    <g
      class="bee-node "
    >
      <circle
        cx="0"
        cy="0"
        fill="none"
        r="78"
        stroke="var(--ph-ok)"
        stroke-dasharray="0"
        stroke-opacity="0.8"
      />
      <ellipse
        cx="0"
        cy="-36"
        fill="var(--ph-node-fill)"
        rx="70"
        ry="8.64"
        stroke="var(--ph-node-stroke)"
      />
      <rect
        fill="var(--ph-node-fill)"
        height="72"
        stroke="var(--ph-node-stroke)"
        width="140"
        x="-70"
        y="-36"
      />
      <ellipse
        cx="0"
        cy="36"
        fill="var(--ph-node-fill)"
        rx="70"
        ry="8.64"
        stroke="var(--ph-node-stroke)"
      />
      <path
        d="M 0,0 m 4.776122516674678e-15,-78 A 78,78 0 0 1 4.897660523286455,-77.84608481740518"
        fill="none"
        stroke="var(--ph-ok)"
        stroke-width="3"
      />
      <g
        fill="var(--ph-text)"
        text-anchor="middle"
      >
        <text
          font-size="12"
          font-weight="700"
          y="-8"
        >
          BeeCounter
        </text>
        <text
          font-size="11"
          y="10"
        >
          alpha
        </text>
      </g>
    </g>
  </svg>
</div>
`;

exports[`BeeNodeShape snapshots > Seeder 1`] = `
<div>
  <svg>
    <g
      class="bee-node "
    >
      <circle
        cx="0"
        cy="0"
        fill="none"
        r="78"
        stroke="var(--ph-ok)"
        stroke-dasharray="0"
        stroke-opacity="0.8"
      />
      <path
        d="M -60.62,-36 L 60.62,-36 L 70,0 L 60.62,36 L -60.62,36 L -70,0 Z"
        fill="var(--ph-node-fill)"
        stroke="var(--ph-node-stroke)"
      />
      <g
        fill="var(--ph-text)"
        text-anchor="middle"
      >
        <text
          font-size="12"
          font-weight="700"
          y="-8"
        >
          Seeder
        </text>
        <text
          font-size="11"
          y="10"
        >
          alpha
        </text>
      </g>
    </g>
  </svg>
</div>
`;

exports[`BeeNodeShape snapshots > Trigger 1`] = `
<div>
  <svg>
    <g
      class="bee-node "
    >
      <circle
        cx="0"
        cy="0"
        fill="none"
        r="78"
        stroke="var(--ph-ok)"
        stroke-dasharray="0"
        stroke-opacity="0.8"
      />
      <path
        d="M -70,-36 L 70,0 L -70,36 Z"
        fill="var(--ph-node-fill)"
        stroke="var(--ph-node-stroke)"
      />
      <polyline
        fill="none"
        points="-6,-10 4,0 -6,10"
        stroke="var(--ph-node-stroke)"
      />
      <g
        fill="var(--ph-text)"
        text-anchor="middle"
      >
        <text
          font-size="12"
          font-weight="700"
          y="-8"
        >
          Activator
        </text>
        <text
          font-size="11"
          y="10"
        >
          alpha
        </text>
      </g>
    </g>
  </svg>
</div>
`;

exports[`BeeNodeShape snapshots > Worker 1`] = `
<div>
  <svg>
    <g
      class="bee-node "
    >
      <circle
        cx="0"
        cy="0"
        fill="none"
        r="78"
        stroke="var(--ph-ok)"
        stroke-dasharray="0"
        stroke-opacity="0.8"
      />
      <path
        d="
    M -60,-36
    H 60 A 10,10 0 0 1 70,-26
    V 26 A 10,10 0 0 1 60,36
    H -60 A 10,10 0 0 1 -70,26
    V -26 A 10,10 0 0 1 -60,-36
    Z
  "
        fill="var(--ph-node-fill)"
        stroke="var(--ph-node-stroke)"
      />
      <g
        opacity="0.5"
      >
        <polyline
          fill="none"
          points="-20,-6 -8,0 -20,6"
          stroke="var(--ph-node-stroke)"
        />
        <polyline
          fill="none"
          points="0,-6 12,0 0,6"
          stroke="var(--ph-node-stroke)"
        />
      </g>
      <g
        fill="var(--ph-text)"
        text-anchor="middle"
      >
        <text
          font-size="12"
          font-weight="700"
          y="-8"
        >
          Worker
        </text>
        <text
          font-size="11"
          y="10"
        >
          alpha
        </text>
      </g>
    </g>
  </svg>
</div>
`;
